### PR TITLE
Series upgrade can be done in massively parallel batches

### PR DESCRIPTION
**Pull request by ChrisMacNaughton**
Due to the nature of the control plane outage required for a series upgrade, most of the API services can be upgraded massively in parallel. The idea is that it is safe to upgrade things in batches like:

```python
SERVICE_GROUPS = collections.OrderedDict([
    ('Stateful Services', ['percona-cluster', 'rabbitmq-server']),
    ('Core Identity', ['keystone', 'ceph-mon']),
    ('Control Plane', [
        'aodh', 'barbican', 'ceilometer', 'ceph-fs',
        'ceph-radosgw', 'cinder', 'designate',
        'designate-bind', 'glance', 'gnocchi', 'heat', 'manila',
        'manila-generic', 'neutron-api', 'neutron-gateway', 'placement',
        'nova-cloud-controller', 'openstack-dashboard']),
    ('Data Plane', [
        'nova-compute', 'ceph-osd', 'swift-proxy', 'swift-storage'])
])
```

Given the above ordering (Stateful Services, Core Identity, etc), the Control Plane group can series-upgrade every unit in parallel. This can drastically reduce the time to perform a series upgrade!

### CODE REVIEW

**Comment by ChrisMacNaughton** [comment]
Something that's worth discussing is that the current versions of the series upgrade code have `files` and `workaround_script` as options. This version currently doesn't as I couldn't find anything using those bits of code. I may have missed something so I'd love feedback from @thedac about it

File: zaza/openstack/utilities/parallel_series_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+    pause_non_leader_subordinate=True,
+    pre_upgrade_functions=None,
+    post_upgrade_functions=None,
+    post_application_upgrade_functions=None,
+    completed_machines=None,
+    follower_first=False,
+    files=None,
+    workaround_script=None
+):
+    """Perform series upgrade on an application in parallel.
```
Comment (gnuoy) [inline zaza/openstack/utilities/parallel_series_upgrade.py:251]: s/parallel/series/

File: zaza/openstack/utilities/parallel_series_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+    """
+    try:
+        await run_on_machine(machine, 'sudo init 6 & exit')
+        # await run_on_machine(unit, "sudo reboot && exit")
+    except subprocess.CalledProcessError as e:
+        logging.warn("Error doing reboot: {}".format(e))
+        pass
+
+
+async def run_on_machine(machine, command, model_name=None, timeout=None):
```
Comment (gnuoy) [inline zaza/openstack/utilities/parallel_series_upgrade.py:540]: Is there not a method in zaza.model that does this already?

File: zaza/openstack/utilities/parallel_series_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+    """
+    logging.info('Upgrading ' + machine)
+    do_release_upgrade_cmd = (
+        'yes | sudo DEBIAN_FRONTEND=noninteractive '
+        'do-release-upgrade -f DistUpgradeViewNonInteractive')
+
+    await run_on_machine(machine, do_release_upgrade_cmd, timeout="120m")
+
+
+# TODO: Move these functions into zaza.model
```
Comment (gnuoy) [inline zaza/openstack/utilities/parallel_series_upgrade.py:600]: +1

File: zaza/openstack/utilities/parallel_series_upgrade.py [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+
+
+def _unit_idle(app, unit_name):
+    async def f():
+        x = await get_agent_status(app, unit_name)
+        return x == "idle"
+    return f
+
+
+async def get_agent_status(app, unit_name):
```
Comment (gnuoy) [inline zaza/openstack/utilities/parallel_series_upgrade.py:628]: Seems like another method that could go over to zaza.model

File: zaza/openstack/utilities/parallel_series_upgrade.py (hunk 1 above)
Comment (ChrisMacNaughton) [reply zaza/openstack/utilities/parallel_series_upgrade.py:251]: This is performing a series upgrade, in parallel

File: zaza/openstack/utilities/parallel_series_upgrade.py (hunk 2 above)
Comment (ChrisMacNaughton) [reply zaza/openstack/utilities/parallel_series_upgrade.py:540]: Currently, nope; there's just `run_on_unit` and `run_on_leader`, both of which expect an actual application unit name

**Pull request merged**